### ISSUE TITLE
RequiresMountsFor= Documentation

### ISSUE THREAD

**Issue by Tensho** [systemd/systemd#12257]
https://www.freedesktop.org/software/systemd/man/systemd.unit.html#RequiresMountsFor=

The current documentation is not intuitive regarding the fact an OS administrator has to provide mount unit aside of the unit she places `RequiresMountsFor` directive. Documentation states `systemd` takes all the heavy lifting to create mount unit and setup required dependencies for 
 **only all mount units**. I guess a new guy in `systemd` like me doesn't realise instantly the need for providing mount unit file explicitly, despite such unit is created by systemd. My assumption is `systemd` waits for some notification from OS about the mount event and then run a dependent service.

In a nutshell – add a paragraph to the `RequiresMountsFor` doc about explicit upfront mount unit file creation. Something like this:

> Systemd requires a corresponding mount unit file to support the unit where `RequiresMountsFor` reference is placed. Systemd doesn't automatically infer the mount points out of its control.

Maybe it's worth to extend the `RequiresMountsFor` directive to support not only the `mount` unit type but either provide the described behaviour for the `service` unit type. What do you think?

**Comment by poettering** [comment]
Hmm? I am not sure I follow. RequiresMountsFor= will add Requires= and After= deps automatically for all mount units in the specified paths. Since systemd automatically synthesizes mount units from /etc/fstab and from all currently established kernel mounts any mount defined should be covered fully and automatically. by the RequiresMountsFor= logic.

Hence, what are you trying to say? what didn't work the way you initially expected it to work?

**Comment by Tensho** [comment]
Thank you for considering this request 🙇 I'm pretty new to `systemd`, so it's reasonable to assume my understanding may be wrong ^_^ However, if I understand it correctly, there are other places to get mount units except of generated from `/etc/fstab` on start. For example, AWS EC2 mounts EBS volume within `cloud-init` out of `systemd` control. Initially I expected `systemd` listens for some kind of mount events (guess `inotify`), compares them to what is stated in `RequiresMountsFor=` and creates a brand new mount unit. Only after that, the service with `RequiresMountsFor=` declaration starts. In the real world I see `systemd` does infer new moun unit from `RequiresMountsFor=` without expected behaviour.

Let me give you example for Nginx. Here is a systemd unit file from the package:

```
# /etc/systemd/system/nginx.service

[Unit]
Description=A high performance web server and a reverse proxy server
After=network.target

[Service]
Type=forking
PIDFile=/run/nginx.pid
ExecStartPre=/usr/sbin/nginx -t -q -g 'daemon on; master_process on;'
ExecStart=/usr/sbin/nginx -g 'daemon on; master_process on;'
ExecReload=/usr/sbin/nginx -g 'daemon on; master_process on;' -s reload
ExecStop=-/sbin/start-stop-daemon --quiet --stop --retry QUIT/5 --pidfile /run/nginx.pid
TimeoutStopSec=5
KillMode=mixed

[Install]
WantedBy=multi-user.target
```
Here is a drop-in with `RequiresMountsFor=` directive:
```
# /etc/systemd/system/nginx.service.d/override.conf

[Unit]
RequiresMountsFor=/application

[Service]
Restart=always
```
These files are not enough to be sure Nginx will run after `/application` mount. I have to provide the mount unit explicitly to make it work:
```
[Unit]
Description=Mount EBS volume at /application

[Mount]
DirectoryMode=0755
Type=ext4
What=/dev/nvme1n1
Where=/application
```

**Comment by poettering** [comment]
not sure i follow. but of course if you want something to be mounted at boot, then you either need to provide an fstab entry for it, or write a .mount unit. A RequiresMountsFor= line alone is not sufficient, I mean where would it know the mount options, source device or file system from?

**Comment by poettering** [comment]
systemd synthesizes .mount units automatically from /etc/fstab entries as well as all mounts established at runtime by the user in the kernel. Thus to make a .mount unit appear in systemd you have three options:

1. write a native .mount unit for it
2. write an /etc/fstab entry
3. just mount it manually using the "mount" command

In the cases 1 + 2 above systemd is able to mount the mount for you, and RequiresMountsFor= hence particularly makes sense. In the case 3 systemd just tracks the mounts already established, i.e. operate after-the-fact and thus RequiresMountsFor= is less useful.

Anyway, I am not sure where the confusion comes from?

This behaviour is documented in https://www.freedesktop.org/software/systemd/man/systemd.mount.html# btw, I'd really suggest reading through that man page first?

**Comment by Tensho** [comment]
> Mount points created at runtime (independently of unit files or /etc/fstab) will be monitored by systemd and appear like any other mount unit in systemd. See /proc/self/mountinfo description in proc(5).

I guess this is the missing piece I have overlooked. For my example, user data shell script (I haven't provided it above) creates mount point. Then `systemd` notices that by `/proc/self/mountinfo` change and creates `application.mount` unit. There is no primary connection between `RequiresMountsFor=` and mount unit creation. As stated in `RequiresMountsFor=` it only sets dependency, nothing more. Sorry I confused you, Lennart 🙇 

Thank you very much for the thorough explanation 🙇

**Comment by carmineesposito73** [comment]
I have a question:
```
[Unit]
Description=my-backup - A simple backup service using tar
After=data.mount dbus.service
RequiresMountsFor=/data /home

[Service]
Type=oneshot
RemainAfterExit=true
User=carmine
ExecStart=/bin/bash -c 'echo my-backup-start: $(date) > /home/carmine/my-services.txt'
ExecStop=/usr/local/bin/my-backup.sh -r 7 -s '/home/carmine/*' \
    -l homedir -d /data/backup/carmine

[Install]
WantedBy=multi-user.target
```

If I stop the service, the script works, but during shutdown it fails because `/data` is unmounted before the my-backup service is stopped.

```
### BOOT
Dec 17 12:10:08 X-cooler systemd[1]: Mounting /data...
Dec 17 12:10:08 X-cooler systemd[1]: Mounted /data.
Dec 17 12:10:10 X-cooler systemd[1]: Starting my-backup - A simple backup service using tar...
Dec 17 12:10:12 X-cooler systemd[1]: Finished my-backup - A simple backup service using tar.
. . .
### SHUTDOWN
Dec 17 12:16:56 X-cooler systemd[1]: data.mount: Succeeded.
Dec 17 12:16:56 X-cooler systemd[1]: Unmounted /data.
Dec 17 12:16:57 X-cooler systemd[1]: Stopping my-backup - A simple backup service using tar...
Dec 17 12:16:57 X-cooler my-backup.sh[2587]: ############################################
Dec 17 12:16:57 X-cooler my-backup.sh[2587]: # my-backup.sh - script to backup my files #
Dec 17 12:16:57 X-cooler my-backup.sh[2587]: ############################################
. . .
Dec 17 12:16:57 X-cooler my-backup.sh[2678]: ERROR: Destination directory /data/backup/carmine not found
Dec 17 12:16:57 X-cooler systemd[1]: my-backup.service: Control process exited, code=exited, status=3/NOTIMPLEMENTED
Dec 17 12:16:57 X-cooler systemd[1]: my-backup.service: Failed with result 'exit-code'.
Dec 17 12:16:57 X-cooler systemd[1]: Stopped my-backup - A simple backup service using tar.
```
Should `RequiresMountsFor` work on shutdown? Do I miss something?

**Comment by kevinh-canva** [comment]
Hello, I have the exact same use case as @carmine73 to run an `ExecStop` at backup, and my service also has a `RequiresMountsFor` on an external EBS volume.

While this sometimes still work, I have encountered a couple of times where the volume is unmounted before `ExecStop` finished. See the logs below:

```
devbox-validate.amazon-ebs.stable-edm: Nov 02 14:55:05 ip-10-20-10-49 systemd[1]: Stopping Setup devbox for coder...
    devbox-validate.amazon-ebs.stable-edm: ░░ Subject: A stop job for unit devbox@coder.service has begun execution
    devbox-validate.amazon-ebs.stable-edm: ░░ Defined-By: systemd
    devbox-validate.amazon-ebs.stable-edm: ░░ Support: http://www.ubuntu.com/support
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ A stop job for unit devbox@coder.service has begun execution.
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ The job identifier is 17157.
    devbox-validate.amazon-ebs.stable-edm: Nov 02 14:55:05 ip-10-20-10-49 sync.sh[14280]: ERROR: Unable to create local directories(/home/coder/.ansible/tmp): [Errno 13] Permission denied: b'/home/coder/.ansible'
    devbox-validate.amazon-ebs.stable-edm: Nov 02 14:55:05 ip-10-20-10-49 systemd[1]: devbox@coder.service: Control process exited, code=exited, status=5/NOTINSTALLED
    devbox-validate.amazon-ebs.stable-edm: ░░ Subject: Unit process exited
    devbox-validate.amazon-ebs.stable-edm: ░░ Defined-By: systemd
    devbox-validate.amazon-ebs.stable-edm: ░░ Support: http://www.ubuntu.com/support
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ An ExecStop= process belonging to unit devbox@coder.service has exited.
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ The process' exit code is 'exited' and its exit status is 5.
```

```
devbox-validate.amazon-ebs.stable-edm: Nov 02 14:55:05 ip-10-20-10-49 systemd[1]: home-coder.mount: Deactivated successfully.
    devbox-validate.amazon-ebs.stable-edm: ░░ Subject: Unit succeeded
    devbox-validate.amazon-ebs.stable-edm: ░░ Defined-By: systemd
    devbox-validate.amazon-ebs.stable-edm: ░░ Support: http://www.ubuntu.com/support
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ The unit home-coder.mount has successfully entered the 'dead' state.
    devbox-validate.amazon-ebs.stable-edm: Nov 02 14:55:05 ip-10-20-10-49 systemd[1]: Unmounted /home/coder.
    devbox-validate.amazon-ebs.stable-edm: ░░ Subject: A stop job for unit home-coder.mount has finished
    devbox-validate.amazon-ebs.stable-edm: ░░ Defined-By: systemd
    devbox-validate.amazon-ebs.stable-edm: ░░ Support: http://www.ubuntu.com/support
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ A stop job for unit home-coder.mount has finished.
    devbox-validate.amazon-ebs.stable-edm: ░░
    devbox-validate.amazon-ebs.stable-edm: ░░ The job identifier is 17277 and the job result is done.
```

`devbox@coder.service` has a `RequiresMountsFor=/home/coder` that is not honoured during shutdown

```
[Unit]
RequiresMountsFor=/home/%i
```

The volume is unmounted before ExecStop has a chance to execute ansible, and since the underlying directory at `/home/coder` is owned by root, it results in a permission denied error when ansible (running as `coder`) is trying to create a subdirectory there.
